### ISSUE TITLE
feature request ffbinaries.findBinary()

### ISSUE THREAD

**Issue by kevinGodell** [vot/ffbinaries-node#15]
I was just thinking about how i am using your module. Before starting a download of ffmpeg, I first check to see if there is a copy in the current directory(and check if it is executable). If not found there, I run a command to see if it is in PATH by calling `ffmpeg -version` and parsing the output to get the version number. If not found there, I then use your module to install fresh copy.

It could be a good addition if you incorporated that utility function to check for existence of binary.
```
const binary = ffbinaries.findBinary('ffmpeg');

//returns null if not found in the usual location, current dir, home dir, PATH, ...
if (binary) {
    console.log(binary.path, binary.version, ...etc)
}
```

I kinda do this starting about [here](https://github.com/kevinGodell/rtsp-cam-tester/blob/master/lib/configure.js#L15). I also have to check `process.pkg` because I am packaging it into executable using nodejs module "pkg" and it does not give the correct running cwd when using process.cwd().

**Comment by kevinGodell** [comment]
I had another thought to expand on the previous request. A directory hint could be given for edge cases like mine where directory may not be in a regular location.
`const binary = ffbinaries.findBinary('ffmpeg', 'hint_to_directory_location');`

**Comment by vot** [comment]
Hi Kevin,

I actually fixed some of this yesterday in version 1.0.5 :)
Whenever you execute `ffbinaries.downloadFiles` it will check if the file already exists (it was supposed to do that already but I was checking for the filename of the zip file instead of actual binary).
That means you don't have to check for binary in the local folder anymore :)

I would like to add a method to check if the binary is installed globally on the system as well (I was thinking about using https://www.npmjs.com/package/which but will have to check it on Windows - I think it might be a bit more thorough and check multiple paths rather than the one where node binary resides). I guess that'd cover most use cases.

Potentially might be worth adding the option to install it globally and symlink from the global path.

**Comment by kevinGodell** [comment]
I had to take a different route because I wanted to check if ffmpeg exists and is executable *first*, and then if not found I **prompt** the user so he has a choice to install it or not. So, in my situation I didn't want to install something without direct user permission(button press).

My script, as it is, works on mac, linux, and windows to check if ffmpeg exists in local folder(and checks if executable and runs chmod +x if not executable). If not found there, it makes a call to it globally to see if i can get the version number with `ffmpegVersion = execSync(`${ffmpegPath} -version`).toString().split(' ')[2];` wrapped in a try catch.

I just looked at the source code for "which" and they don't seem to check executable permissions on file, but maybe i missed it.

This was only an idea I thought might give some added functionality, but it is not absolutely necessary. I am working on the implementation with my small app and I think my guy who is running a much larger app with a growing user base may implement my technique for installing ffmpeg with your module.

**Comment by vot** [comment]
Fair points and I do like the idea of just allowing people to check if it's already there.

It might not always be to just automatically jump to the download and the version check seems useful for the sake of more flexible workflows.
Not sure if automatically setting the permissions would be a reasonable thing to do automatically (especially in system directories) but that could be an opt-in.

I'm going to start v1.0.6 branch.

**Comment by vot** [comment]
OK so this is more or less what I have in mind:
https://github.com/vot/ffbinaries-node/compare/master...v1.0.6

I borrowed a lot of your code but skipped chmod for now because of Windows - I'm guessing files are always executable there? Would we ever run into trying to execute `chmod` on Windows?

Other than chmod do you see anything missing there?
I'll try to add some reasonable tests shortly as well before merging it in and give sync/async aspect of it some thought.

**Comment by kevinGodell** [comment]
Pretty cool. The thing about chmod on windows is that it would never have been called since it return true(or maybe simple doesnt throw error) when calling fs.accessSync(ffmpegLocalPath, fs.constants.X_OK); I guess i should have excluded that possibility on windows since it is easy enough to check when we are on windows. Will your implementation return on the first binary found?

**Comment by vot** [comment]
Yeah seems like there's just no notion of executable flag in Windows so nothing to worry about, it will always return true.

In that case wrapping it in the `if` block [here](https://github.com/vot/ffbinaries-node/compare/master...v1.0.6#diff-49c3a8fa80ea82c103ccac5453a3dc33R398) will be sufficient - Windows will just never step into that condition.

---

I'll add `chmod +x`, some tests and documentation this week (hopefully in the next day or two).
Really coming back to the idea of making this async - would you rather have it async, sync or both?

And yes, it does currently return on the first hit (meaning anything in `PATH` env var will be checked before the `additionalPaths`). Do you think you'd possibly want to see more than one hit, perhaps if there are different versions in different locations?

**Comment by kevinGodell** [comment]
I had planned my implementation slightly different so that somebody can just drop a newer copy of ffmpeg in the working directory to override whatever other ffmpeg that may already be installed and in PATH. Personally, I would prefer to look at the hinted location first and fall back to checking PATH. Think about it, wouldn't it be nice to try out the latest version of ffmpeg by just dropping it to the directory without having to mess with PATH vars?

For me, sync would work best because I am running it at startup to make the necessary checks and set a flag indicating whether or not it was found.

Returning on the first hit is fine, I was just curious. That seems like a good option instead of returning array and then end user has to pick which one to use.

**Comment by kevinGodell** [comment]
Looks like it is coming along good. I haven't had time to test that branch, but it looks like it will provide nice added features. Thanks.

**Comment by vot** [comment]
No worries, cheers for a good suggestion :+1:

I still didn't get around to writing tests for this but hoping to release it through npm at some point this week - will tag you in a response here once it's ready.

**Comment by vot** [comment]
Hi @kevinGodell, I actually managed to write some tests today.

Version 1.0.6 has now been published to npm.

You should be able to replace the whole `configure.js` with this one-liner: `ffbinaries.locateBinariesSync('ffmpeg', {paths: [process.cwd()], ensureExecutable: true })` (...and a few variable assignments).

**Comment by kevinGodell** [comment]
hi. I just had a chance to try it. For locateBinariesSync, seems like 'components' and 'opts' required. When I first tried it, I forgot to pass options and it threw error when checking properties of undefined object. My bad. I wasn't sure if it was mandatory or optional for it to receive options.

Also, when it was trying to get existing ffmpeg version, it does a split by '-' which took my global ffmpeg version git-2017-11-02-2805c8d and just spit out 'git' as version.

That was just at first glance, I am going to implement it now and test functionality. Thanks.

**Comment by kevinGodell** [comment]
Looking good. I think I have it fully implemented to take advantage of your new features. 

One last thing, when I call the locateBinariesSync method and ffmpeg is not installed yet ,  "chmod: cannot access 'null': No such file or directory" is printed to console in mac and ubuntu. I think that call to chmod could use a null check on the path being sent to it first. Otherwise, thanks for all of your hard work.

**Comment by kevinGodell** [comment]
Just tried it on windows 7 x64 and the logic does not seem to detect the system type. Needs a  detectPlatform() at some point so that it can add the .exe for windows. I also tried your test file(04) and it had the same results of not being able to detect ffmpeg binaries. I tried plugging detectPlatform() into a couple different locations and it definitely will fix the binaryFilename. Thanks.

**Comment by vot** [comment]
Hi @kevinGodell,

As usual thanks for the feedback.
I just published 1.0.7 - a few lines of code altogether (I've identified a few things I wanted to correct as well in CLI and tests).

It should address all these issues:
- allow you to omit `opts`
- stop `chmod +x` when the file hasn't been found
- look for the correct filename on windows

Thanks for spotting those, glad to have these weird issues out of the way quickly 👍

**Comment by kevinGodell** [comment]
I just deployed it and everything works great now in linux, mac, and windows. Thank you very much. If you wanted to see it in action, you could test my [binary release](https://github.com/kevinGodell/rtsp-cam-tester/releases). If it already detects the version of ffmpeg on your system when you go to `localhost:8181`, then you can go to `localhost:8181/install`  to force the ffmpeg download from there.

I have a slightly modified version of your module when it comes to getting the version. I don't split it by "-" since some ffmpeg versions that may already exist on the system have that dash in the version name like `git-2017-11-02-2805c8d`. Is there any way you could include another property to return **.fullVersion** so that I can get it without it being split?

**Comment by vot** [comment]
Hi Kevin, I actually meant to remove that extra string parsing in this release.
`version` field will return full version (not splitting by `-`) in 1.0.8, I've got some linting updates coming as well so should be within the next few days.

Will also check out your app as well while I'm at it, haven't been able to get around to it just yet :)

**Comment by kevinGodell** [comment]
Just had a thought for another feature before your next release. Would it be possible to pass a flag to ignore checking for the existence of the binary(and existence of cache) when attempting to initiate downloading? The use case scenario would be in `updating` situations when user wants to get newer version of ffmpeg while already having it on the system. Currently, the download will stop short after if finds binary at the target location.

**Comment by vot** [comment]
Hi @kevinGodell,

I left stricter linting for a future release as I didn't want to introduce any bugs by rushing that.

In the meantime I published v1.0.8 to npm which returns full version in `locateBinariesSync` and adds `force` option to `downloadFiles` / `downloadBinaries` method.

Setting `force` to `true` will still use cache (there's `clearCache` method to get rid of that if needed) but it will overwrite any file in the specified destination.

Many thanks for your suggestions - I'm sure they'll come in handy for many people in their workflows.

**Comment by kevinGodell** [comment]
I just included your latest version in my deployment. Thanks for all of the features! I don't have any more ideas, so I guess you can close this feature request issue. Thanks again.

**Comment by vot** [comment]
Perfect, glad I could help :) Will close the request now.

**Comment by kevinGodell** [comment]
One last comment that did not seem worthy of opening a new issue. I noticed something that does not affect performance, but is more of an aesthetics things. On windows, when returning the path to a found binary, there will be a mix of backslashes and a forward slash, `L:\WebstormProjects\ffmpeg-streamer/ffmpeg.exe`. Wrapping the binaries path with path.resolve before returning it fixes up those slashes. Like I said, it doesn't affect performance, just looks a little funny to me. Feel free to ignore this NON-issue. Thanks again.

**Comment by vot** [comment]
Thanks for bringing this up - will include a fix for this in the linting work, might as well be consistent :)

BTW I looked at your app the other day, really cool stuff - I wasn't even aware of `pkg` at all, that's some really cool process you've assembled there. Can't wait to find a reason for me to use it in one of my projects ;)
TBH I wanted to put together something like that myself at one point, I might contribute a bit to your project when I get some spare time - it's a great idea that hasn't really been done well in the past.
